src/feature/nodelist/dirlist.c

```diff
@@ -49,6 +49,31 @@ static smartlist_t *trusted_dir_servers = NULL;
  * and all fallback directory servers. */
 static smartlist_t *fallback_dir_servers = NULL;
 
+/** Return true iff the given v4 address is the one of a configured directory
+ * authority matching the given type.
+ *
+ * This needs to be fast because it can be called at each directory connection
+ * in order to learn if the request is from an authority or not. */
+bool
+dirlist_addr_is_trusted_dir_by_type(const tor_addr_t *addr,
+                                    const dirinfo_type_t type)
+{
+  if (BUG(!addr) || BUG(!trusted_dir_servers)) {
+    return false;
+  }
+
+  /* Go over all trusted directories. We have very few of these so this should
+   * be fairly fast. In the future, if this shows up in the profile, we can
+   * move to have an address set instead. */
+  SMARTLIST_FOREACH_BEGIN(trusted_dir_servers, const dir_server_t *, ds) {
+    if ((ds->type & type) && tor_addr_eq_ipv4h(addr, ds->addr)) {
```

Comment:
Please check for IPv6 addresses.

It won't cost much extra CPU time, because address types are checked first, and the volume of IPv6 bugs is low. But it will avoid weird bugs in future, when we reach IPv6 feature parity.

Reply:
If the incoming address is an IPv6 address, you'll want to check dir_server_t's ipv6_addr:
https://github.com/torproject/tor/blob/4f02812242d1fd90d859eb98ac3fb1ed182f18cf/src/feature/dirclient/dir_server_st.h#L27

Reply:
I think these issues are ok, but they need to be talked about in the function comments, in case they cause bugs in future:

You're checking the hard-coded (or configured) list of directory authorities, rather than the current consensus. That's probably ok, but it requires the directory authorities to update their configurations when the set of authority addresses changes. (Usually, if authorities advertise a new address in their descriptors, it gets picked up in the consensus, and used by other directory authorities, without a local config change. But this function is an exception to that rule.)

And you're checking the inbound ORPort address, rather than the OutboundBindAddress. How many directory authorities have different inbound and outbound addresses?